README.md
# uxcore-pagination

pagination ui component for react  
[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url] 
[![NPM downloads][downloads-image]][npm-url]
[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-pagination.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-pagination
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-pagination.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-pagination
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-pagination.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-pagination?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-pagination.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-pagination
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-pagination.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-pagination#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-pagination.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-pagination.svg
[sauce-url]: https://saucelabs.com/u/uxcore-pagination

#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-pagination
$ cd uxcore-pagination
$ npm install
$ npm start
```

## Usage

```js
var Pagination = require('uxcore-pagination');
React.render(
	<Pagination onChange={onChange} total={50} />, document.getElementById('target'));
```

### demo
http://uxcore.github.io/uxcore/components/pagination/

## API

## Props

|参数|说明|类型|默认值|
|---|----|---|------|
|locale|语言(zh-cn/en-us)|string|zh-cn|
|current|当前页数|number|1|
|total|数据总数|number/jsx|0|
|pageSize|每页条数|number|10|
|onChange|页码改变的回调，参数是改变后的页码|function|noop|
|showTotal|是否显示共多少条|boolean|false|
|showQuickJumper|是否可以快速跳转至某页|bool|false|
|showSizeChanger|是否可以改变 pageSize|bool|false|
|sizeOptions|sizeChanger 显示的可选 pageSize|array|[10, 20, 30, 40]|
|onShowSizeChange|pageSize 变化的回调|function|noop|
|className|当为「mini」时，是小尺寸分页|string||
|simple|当添加该属性时，显示为简单分页|object|无|
|maxUnfoldedLength|页码显示的数量，更多的使用...代替|Number|9|
|getSelectPopupContainer|指定下拉选择框渲染的容器|function():HTML Element|插在 body 下的一个 div|


HISTORY.md
# History

---
## 0.8.5
* `NEW` 新增 locale key en_US zh_CN (#35)

## 0.8.4
* `NEW` support localePack

## 0.8.3
* `NEW` zh-hk support

## 0.8.2

* `NEW` add `totalSizeOffset` Props API

## 0.8.0

* `CHANGED` change all icons to uxcore-icon

## 0.7.9

* `NEW` add `totalSizeOffset` Props API

## 0.7.7

* `CHANGED` change all icons to uxcore-icon

## 0.7.8

`FIX` roll-back to version 0.7.6 because of legao style problem

## 0.7.6

* `CHANGED` add `maxUnfoldedLength` Props API

## 0.7.2

* `CHANGED` support js style export

## 0.7.1

* `CHANGED` modify i18n text

## 0.7.0

* `CHANGED` Migrate to react 16.x

## 0.6.2

* `CHANGED` Migrate to react 15.x

## 0.6.1

* `CHANGED` improve style when total count is large

## 0.6.0

* `CHANGED` add quick jumper button
* `CHANGED` quick jumper input value follow the current page

## 0.5.2

* `CHANGED` add i18n for unknown total
* `FIXED` incorrect style

## 0.5.0

`CHANGED` adjust quick jumper position

## 0.4.1

`CHANGED` style change

## 0.4.0

* `CHANGED` new style
* `NEW` add new style for unknown total

## 0.3.1

`NEW` add new Prop `getSelectPopupContainer`

## 0.3.0

`CHANGED` update `uxcore-select` to ~0.4.0

## 0.2.4

`CHANGED` update dependency `uxcore-select` to ~0.3.0

## 0.2.3

`FIXED` fix pagination size changer style bug
`CHANGED` new scaffold

## 0.2.2

`CHANGED` update uxcore-select2 to ~0.2.0

## 0.2.0

`NEW` new style

## 0.1.20

`NEW` internationalization, now support 'zh-cn' & 'en-us'

## 0.1.9
`CHANGED` update scaffold

## 0.1.8
`CHANGED` change arrow with kuma-icon

## 0.1.7

`FIXED` fix issue #6

## 0.1.6

`FIXED` fix issue #7

## 0.1.5

`CHANGED` change css prefix

## 0.1.4

`CHANGED` upgrade to react@0.14


## 0.1.3

`CHANGED` Forked from rc-pagination
`CHNAGED` simple style changed


tests/Pagination.spec.js
import expect from 'expect.js';
import React from 'react';
import ReactDOM from 'react-dom';
import assign from 'object-assign';
import TestUtils, { Simulate } from 'react-addons-test-utils';
import { mount } from 'enzyme';
import Pagination from '../src';

function renderPageWithProps(pageProps) {
  const props = {
    total: 50,
    pageSize: 10,
    className: 'test-page',
    showTotal: true,
    showQuickJumper: true,
    showSizeChanger: true,
  };
  assign(props, pageProps);
  const wrapper = mount(<Pagination {...props} />);
  return wrapper;
}

describe('Pagination', () => {
  describe('render', () => {

    it('should render correctly', (done) => {
      const wrapper = renderPageWithProps();
      expect(wrapper.find('.test-page').length).to.be(1);
      done();
    });
  });

  describe('control', () => {
    const wrapper = renderPageWithProps({
      total: 500,
      pageSize: 10,
    });
    it('should set page correctly', (done) => {
      expect(wrapper.node.state.current).to.be(1);
      wrapper.node._handleChange(2)
      expect(wrapper.node.state.current).to.be(2);
      done();
      wrapper.node._next();
      expect(wrapper.node.state.current).to.be(3);
      wrapper.node._prev();
      expect(wrapper.node.state.current).to.be(2);
      wrapper.node._jumpNext();
      expect(wrapper.node.state.current).to.be(7);
      wrapper.node._jumpPrev();
      expect(wrapper.node.state.current).to.be(2);
      wrapper.node._handleChange(30);
      wrapper.node.options._changeSize(20);
      expect(wrapper.node.state.current).to.be(25);
      done();
    });
    it('should quick jumper work correctly', (done) => {
      const input = wrapper.find('.kuma-page-options-quick-jumper > input');
      input.node.value = 10;
      input.simulate('change')
      input.simulate('keyup', {
        keyCode: 13,
      });
      expect(wrapper.node.state.current).to.be(10);
      done();
    });

  });
  

});

src/Options.js
const React = require('react'); 
const ReactDOM = require('react-dom');
const KEYCODE = require('./KeyCode');
const i18n = require('./locale');

class Options extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      current: props.current,
      _current: props.current,
    };

    ['_handleChange', '_changeSize', '_go'].forEach((method) => this[method] = this[method].bind(this));
  }
  render() {
    const props = this.props;
    const state = this.state;
    const prefixCls = `${props.rootPrefixCls}-options`;
    const sizeOptions = props.sizeOptions;
    const pageSize = props.pageSize;
    const changeSize = props.changeSize;
    const quickGo = props.quickGo;
    const Select = props.selectComponentClass;
    let changeSelect = null;
    let goInput = null;

    if (!(changeSize || quickGo)) {
      return null;
    }

    if (changeSize && Select) {
      const Option = Select.Option;
      changeSelect = (
        <Select
          prefixCls={props.selectPrefixCls} showSearch={false}
          className={`${prefixCls}-size-changer`}
          getPopupContainer={props.getPopupContainer}
          optionLabelProp="children"
          dropdownClassName={`${prefixCls}-size-changer-dropdown`}
          defaultValue={sizeOptions.indexOf(pageSize) == -1 ? sizeOptions[0]+"" : pageSize+"" } 
          onChange={this._changeSize}>
          {sizeOptions.map((option, index) => {
            return <Option key={option} value={option + ""}>{option + i18n[props.locale]['items_per_page']}</Option>
          })}
       </Select>
      );
    }

    if (quickGo) {
      goInput = (
        <div title="Quick jump to page" className={`${prefixCls}-quick-jumper`}>
          {i18n[props.locale]['jump_to']}
          <input type="text" className="kuma-input" value={state._current} onChange={this._handleChange.bind(this)} onKeyUp={this._go.bind(this)}/>
          {i18n[props.locale]['page']}
        </div>
      );
    }

    return (
      <div className={`${prefixCls}`}>
        {changeSelect}
        {goInput}
      </div>
    );
  }

  _changeSize(value) {
    this.props.changeSize(Number(value));
  }

  _handleChange(evt) {
    const _val = evt.target.value;

    this.setState({
      _current: _val,
    });
  }

  _go(e) {
    const _val = e.target.value;
    if (_val === '') {
      return;
    }
    let val = Number(this.state._current);
    if (isNaN(val)) {
      val = this.state.current;
    }
    if (e.keyCode === KEYCODE.ENTER) {
      const c = this.props.quickGo(val);
      this.setState({
        _current: c,
        current: c,
      });
    }
  }
}

Options.propTypes = {
  changeSize: React.PropTypes.func,
  quickGo: React.PropTypes.func,
  selectComponentClass: React.PropTypes.func,
  current: React.PropTypes.number,
};

module.exports = Options;

demo/PaginationDemo.js
import React from 'react';
import Pagination from '../src/index';

export default class Demo extends React.Component {

	constructor(props) {
		super(props);
		this.state = {
			index: 1
		}
	}

	handleChange(page) {
		console.log(page);
		this.setState({
			index: page
		})
	}

	render(){
		let me = this;
		return (
			<div>
				<h2>基础分页:</h2>
				<Pagination 
					showSizeChanger
					onChange={onChange} 
					showQuickJumper
					total={50} 
					pageSize={8} 
					locale='en' 
					getSelectPopupContainer={() => {
						const div = document.createElement('div');
						div.className = 'uxcore';
						document.body.appendChild(div);
						return div;
					}} 
				/>
				<h2>更多分页:</h2>
				<Pagination onChange={onChange} total={500} />
				<h2>快速跳转到某一页:</h2>
				<Pagination showQuickJumper={true} onChange={onChange} total={500}/>
				<h2>迷你版本:</h2>
				<Pagination className="mini" current={this.state.index} showTotal={true} showQuickJumper={true} onChange={me.handleChange.bind(me)} total={50} sizeOptions={[5, 10, 20]} pageSize={5} showSizeChanger={true} />
				<h2>简单地翻页:</h2>
				<Pagination simple onChange={me.handleChange.bind(me)} total={50} pageSize={1} />
				<h2>总数未知</h2>
				<Pagination simple onChange={me.handleChange.bind(me)} pageSize={1} />
			</div>
		);
	}
}

function onChange(key){
	console.log(key);
}


src/Pagination.js
/**
 * Forked from project rc-pagination
 * @maintainer eternalsky
 */

const Select = require('uxcore-select2');
const Pager = require('./Pager');
const Options = require('./Options');
const KEYCODE = require('./KeyCode');
const React = require('react'); 
const ReactDOM = require('react-dom');
const i18n = require('./locale');

function noop() {
}

class Pagination extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      current: props.current,
      _current: props.current,
      pageSize: props.pageSize,
    };

    [
      'render',
      '_handleChange',
      // '_handleKeyUp',
      // '_handleKeyDown',
      '_changePageSize',
      '_isValid',
      '_prev',
      '_next',
      '_hasPrev',
      '_hasNext',
      '_jumpPrev',
      '_jumpNext',
    ].forEach((method) => this[method] = this[method].bind(this));
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.current != this.props.current) {
      this.setState({
        current: nextProps.current,
        _current: nextProps.current
      });
    }

    if (nextProps.pageSize != this.props.pageSize) {
      this.setState({
        pageSize: nextProps.pageSize,
      });
    }
  }

  renderTotal() {
    let prefix = this.props.locale == 'zh-cn' ? "共" : ""; 
    if (this.props.showTotal) {
      return <li className={this.props.prefixCls + "-total"}>{prefix + this.props.total + i18n[this.props.locale]['item']}</li>
    }
  }

  render() {
    const props = this.props;

    const prefixCls = props.prefixCls;
    const allPages = this._calcPage();
    const pagerList = [];
    let jumpPrev = null;
    let jumpNext = null;
    let firstPager = null;
    let lastPager = null;

    if ([0, undefined, null].indexOf(props.total) !== -1) {
      return (
        <ul className={`${prefixCls} ${props.className}`}>
          <li title="Previous Page" onClick={this._prev} className={(this._hasPrev() ? '' : `${prefixCls}-disabled `) + `${prefixCls}-prev`}>
            <a className="kuma-icon kuma-icon-chevron-left"></a>
          </li>
          <div title={`Page ${this.state.current}`} className={`${prefixCls}-unknown-total`}>
            <span className={`${prefixCls}-current`}>{this.state._current}</span>
          </div>
          <li title="Next Page" onClick={this._next} className={(this._hasNext() ? '' : `${prefixCls}-disabled `) + `${prefixCls}-next`}>
            <a className="kuma-icon kuma-icon-chevron-right"></a>
          </li>
        </ul>
      )
    }

    if (props.simple) {
      return (
        <ul className={`${prefixCls} ${prefixCls}-simple ${props.className}`}>
          <li title="Previous Page" onClick={this._prev} className={(this._hasPrev() ? '' : `${prefixCls}-disabled `) + `${prefixCls}-prev`}>
            <a className="kuma-icon kuma-icon-chevron-left"></a>
          </li>
          <div title={`Page ${this.state.current} of ${allPages}`} className={`${prefixCls}-simple-pager`}>
            <span className={`${prefixCls}-current`}>{this.state._current}</span>
            <span className={`${prefixCls}-slash`}>/</span>
            {allPages}
          </div>
          <li title="Next Page" onClick={this._next} className={(this._hasNext() ? '' : `${prefixCls}-disabled `) + `${prefixCls}-next`}>
            <a className="kuma-icon kuma-icon-chevron-right"></a>
          </li>
        </ul>
      );
    }

    if (allPages <= 9) {
      for (let i = 1; i <= allPages; i++) {
        const active = this.state.current === i;
        pagerList.push(<Pager rootPrefixCls={prefixCls} onClick={this._handleChange.bind(this, i)} key={i} page={i} active={active} />);
      }
    } else {
      jumpPrev = (<li title="Previous 5 Page" key="prev" onClick={this._jumpPrev} className={`${prefixCls}-jump-prev`}>
        <a></a>
      </li>);
      jumpNext = (<li title="Next 5 Page" key="next" onClick={this._jumpNext} className={`${prefixCls}-jump-next`}>
        <a></a>
      </li>);
      lastPager = <Pager last={true} rootPrefixCls={prefixCls} onClick={this._handleChange.bind(this, allPages)} key={allPages} page={allPages} active={false} />;
      firstPager = <Pager rootPrefixCls={prefixCls} onClick={this._handleChange.bind(this, 1)} key={1} page={1} active={false} />;

      const current = this.state.current;

      let left = Math.max(1, current - 2);
      let right = Math.min(current + 2, allPages);

      if (current - 1 <= 2) {
        right = 1 + 4;
      }

      if (allPages - current <= 2) {
        left = allPages - 4;
      }

      for (let i = left; i <= right; i++) {
        const active = current === i;
        pagerList.push(<Pager rootPrefixCls={prefixCls} onClick={this._handleChange.bind(this, i)} key={i} page={i} active={active} />);
      }

      if (current - 1 >= 4) {
        pagerList.unshift(jumpPrev);
      }
      if (allPages - current >= 4) {
        pagerList.push(jumpNext);
      }

      if (left !== 1) {
        pagerList.unshift(firstPager);
      }
      if (right !== allPages) {
        pagerList.push(lastPager);
      }
    }

    return (
      <ul className={`${prefixCls} ${props.className}`}
        unselectable="unselectable">
        {this.renderTotal()}
        <li title="Previous Page" onClick={this._prev} className={(this._hasPrev() ? '' : `${prefixCls}-disabled `) + `${prefixCls}-prev`}>
          <a className="kuma-icon kuma-icon-chevron-left"></a>
        </li>
        {pagerList}
        <li title="Next Page" onClick={this._next} className={(this._hasNext() ? '' : `${prefixCls}-disabled `) + `${prefixCls}-next`}>
          <a className="kuma-icon kuma-icon-chevron-right"></a>
        </li>
        <Options rootPrefixCls={prefixCls}
          ref={options => this.options = options}
          locale={props.locale}
          selectComponentClass={props.selectComponentClass}
          getPopupContainer={props.getSelectPopupContainer}
          selectPrefixCls={props.selectPrefixCls}
          changeSize={this.props.showSizeChanger ? this._changePageSize.bind(this) : null}
          current={this.state.current}
          pageSize={props.pageSize}
          sizeOptions={props.sizeOptions}
          quickGo={this.props.showQuickJumper ? this._handleChange.bind(this) : null} />
      </ul>
    );
  }

  // private methods

  _calcPage(p) {
    let pageSize = p;
    if (typeof pageSize === 'undefined') {
      pageSize = this.state.pageSize;
    }
    if ([0, undefined, null].indexOf(this.props.total) !== -1) {
      return Infinity;
    }
    return Math.floor((this.props.total - 1) / pageSize) + 1;
  }

  _isValid(page) {
    return typeof page === 'number' && page >= 1 && page !== this.state.current;
  }

  _changePageSize(size) {
    if (typeof size === 'number') {
      let current = this.state.current;

      this.setState({
        pageSize: size,
      });

      if (this.state.current > this._calcPage(size)) {
        current = this._calcPage(size);
        this.setState({
          current: current,
          _current: current,
        });
      }

      this.props.onShowSizeChange(current, size);
    }
  }

  _handleChange(p) {
    let page = p;
    let me = this;
    if (this._isValid(page)) {
      if (page > this._calcPage()) {
        page = this._calcPage();
      }
      this.setState({
        current: page,
        _current: page,
      }, () => {
        me.props.onChange(page);
      });

      return page;
    }

    return this.state.current;
  }

  _prev() {
    if (this._hasPrev()) {
      this._handleChange(this.state.current - 1);
    }
  }

  _next() {
    if (this._hasNext()) {
      this._handleChange(this.state.current + 1);
    }
  }

  _jumpPrev() {
    this._handleChange(Math.max(1, this.state.current - 5));
  }

  _jumpNext() {
    this._handleChange(Math.min(this._calcPage(), this.state.current + 5));
  }

  _hasPrev() {
    return this.state.current > 1;
  }

  _hasNext() {
    return this.state.current < this._calcPage();
  }
}

Pagination.propTypes = {
  current: React.PropTypes.number,
  total: React.PropTypes.number,
  locale: React.PropTypes.string,
  showTotal: React.PropTypes.bool,
  pageSize: React.PropTypes.number,
  sizeOptions: React.PropTypes.array,
  onChange: React.PropTypes.func,
  showSizeChanger: React.PropTypes.bool,
  onShowSizeChange: React.PropTypes.func,
  selectComponentClass: React.PropTypes.func,
  showQuickJumper: React.PropTypes.bool,
};

Pagination.defaultProps = {
  current: 1,
  total: 0,
  locale: 'zh-cn',
  showTotal: false,
  pageSize: 10,
  sizeOptions: [10, 20, 30, 40],
  onChange: noop,
  className: '',
  selectPrefixCls: 'kuma-select2',
  prefixCls: 'kuma-page',
  selectComponentClass: Select,
  showQuickJumper: false,
  showSizeChanger: false,
  onShowSizeChange: noop,
};

Pagination.displayName = 'Pagination';

module.exports = Pagination;
